media-proxy/src/mesh/conn_rdma_tx.cc

```diff
@@ -0,0 +1,180 @@
+
+/**
+ * @brief Handles sending data through RDMA by consuming a buffer, copying data, and transmitting it.
+ * 
+ * This function attempts to consume a pre-allocated buffer from the queue within a specified timeout,
+ * copies the provided data into the buffer, and sends it through the RDMA endpoint. It ensures proper
+ * error handling, retries for buffer availability, and buffer management in case of transmission failure.
+ * 
+ * @param ctx The context for managing the operation.
+ * @param ptr Pointer to the data to be transmitted.
+ * @param sz The size of the data to be transmitted.
+ * @param sent Output parameter indicating the actual number of bytes sent.
+ * @return Result::success on successful transmission, or an appropriate error result on failure.
+ */
+Result RdmaTx::on_receive(context::Context& ctx, void *ptr, uint32_t sz, uint32_t& sent)
+{
+    void *reg_buf = nullptr;
+    constexpr uint32_t TIMEOUT_US = 500000;     // 0.5-second timeout
+    constexpr uint32_t RETRY_INTERVAL_US = 100; // Retry interval of 100 us
+    uint32_t elapsed_time = 0;
+
+    // Attempt to consume a buffer from the queue with a timeout
+    while (elapsed_time < TIMEOUT_US && !ctx.cancelled()) {
+        Result res = consume_from_queue(ctx, &reg_buf);
+        if (res == Result::success) {
+            if (reg_buf != nullptr) {
+                break; // Successfully got a buffer
+            } else {
+                log::debug("Buffer is null, retrying...")(" ", kind2str(_kind));
+            }
+        } else if (res != Result::error_no_buffer) {
+            // Log non-retryable errors and exit
+            log::error("Failed to consume buffer from queue")("result", static_cast<int>(res))
+                      (" ", kind2str(_kind));
+            sent = 0;
+            return res;
+        }
+
+        // Wait before retrying
+        std::this_thread::sleep_for(std::chrono::microseconds(RETRY_INTERVAL_US));
+        elapsed_time += RETRY_INTERVAL_US;
+    }
+
+    // Check if we failed to get a buffer within the timeout
+    if (reg_buf == nullptr) {
+        log::error("Failed to consume buffer within timeout")("timeout_ms", TIMEOUT_US)
+                  (" ", kind2str(_kind));
+        sent = 0;
+        return Result::error_timeout;
+    }
+
+    uint32_t tmp_sent = std::min(static_cast<uint32_t>(trx_sz), sz);
+    if (tmp_sent != trx_sz) {
+        log::debug("Sent size differs from transfer size")("requested_size", tmp_sent)
+                  ("trx_sz", trx_sz)(" ", kind2str(_kind));
+    }
+
+    std::memcpy(reg_buf, ptr, tmp_sent);
+
+    // Transmit the buffer through RDMA
+    int err = libfabric_ep_ops.ep_send_buf(ep_ctx, reg_buf, tmp_sent);
+    notify_buf_available();
+    sent = tmp_sent;
+
+    if (err) {
+        log::error("Failed to send buffer through RDMA")("error", fi_strerror(-err))
+                  (" ", kind2str(_kind));
+
+        // Add the buffer back to the queue in case of failure
+        add_to_queue(reg_buf);
```

Comment:
add_to_queue() returns a return value, check it.

Reply:
fixed in https://github.com/OpenVisualCloud/Media-Communications-Mesh/commit/737483c86a4bf88fdd86b9e223e58acf7f3b7d6d